pkg/engine/lifecycletest/test_plan.go

```diff
@@ -237,6 +265,192 @@ func (op TestOp) runWithContext(
 	return nil, snap, errors.Join(errs...)
 }
 
+// We're just checking that we have the right number of events and
+// that they have the expected types.  We don't do a deep comparison
+// here, because all that matters is that we have the same events in
+// some order.  The non-display tests are responsible for acually
+// checking the events properly.
+func compareEvents(expected, actual []engine.Event) error {
+	encountered := make(map[int]bool)
+	if len(expected) != len(actual) {
+		return fmt.Errorf("expected %d events, got %d", len(expected), len(actual))
+	}
+	for _, e := range expected {
+		found := false
+		for i, a := range actual {
+			if encountered[i] {
+				continue
+			}
+			if a.Type == e.Type {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("expected event %v not found", e)
+		}
+	}
+	return nil
+}
+
+// TODO: dedup with diff_test.go
+func loadEvents(path string) (events []engine.Event, err error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("opening '%v': %w", path, err)
+	}
+	defer contract.IgnoreClose(f)
+
+	dec := json.NewDecoder(f)
+	for {
+		var jsonEvent apitype.EngineEvent
+		if err = dec.Decode(&jsonEvent); err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			return nil, fmt.Errorf("decoding event %d: %w", len(events), err)
+		}
+
+		event, err := bdisplay.ConvertJSONEvent(jsonEvent)
+		if err != nil {
+			return nil, fmt.Errorf("converting event %d: %w", len(events), err)
+		}
+		events = append(events, event)
+	}
+
+	// If there are no events or if the event stream does not terminate with a cancel event,
+	// synthesize one here.
+	if len(events) == 0 || events[len(events)-1].Type != engine.CancelEvent {
+		events = append(events, engine.NewCancelEvent())
+	}
+
+	return events, nil
+}
+
+func assertDisplay(t testing.TB, events []Event, path string) {
+	var expectedStdout []byte
+	var expectedStderr []byte
+	accept := cmdutil.IsTruthy(os.Getenv("PULUMI_ACCEPT"))
+	if !accept {
+		var err error
+		expectedStdout, err = os.ReadFile(path + "/diff.stdout.txt")
```

Comment:
Should this use `filepath.Join` (applies throughout)?

```suggestion
		expectedStdout, err = os.ReadFile(filepath.Join(path, "diff.stdout.txt"))
```

Reply:
It definitely should, fixing that now.